archive/imo/imo1975_q1.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2022 Mantas Bakšys. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mantas Bakšys
+-/
+import data.real.basic
+import data.nat.interval
+import algebra.order.rearrangement
+import algebra.big_operators.ring
+
+/-!
+# IMO 1975 Q1
+
+Let `x₁, x₂, ... , xₙ` and `y₁, y₂, ... , yₙ` be two sequences of real numbers, such that
+`x₁ ≥ x₂ ≥ ... ≥ xₙ` and `y₁ ≥ y₂ ≥ ... ≥ yₙ`. Prove that if `z₁, z₂, ... , zₙ` is any permutation
+of `y₁, y₂, ... , yₙ`, then
+`∑ i in finset.Icc 1 n, (x i - y i) ^ 2 ≤ ∑ i in finset.Icc 1 n, (x i - z i) ^ 2`
+
+# Solution
+
+Firstly, we expand the squares withing both sums and distribute into separate finite sums. Then,
+noting that `∑ i in finset.Icc 1 n, (y i) ^ 2 = ∑ i in finset.Icc 1 n, (z i) ^ 2`, it remains to
+prove that `∑ i in finset.Icc 1 n, x i * z i ≤ ∑ i in finset.Icc 1 n, x i * y i`, which is true by
+the Rearrangement Inequality
+-/
+
+open_locale big_operators
+
+/-- Let `n` be a natural number, `x` and `y` be as in the problem statement and `σ` be the
+permutation of natural numbers such that `z = y ∘ σ` -/
+variables (n : ℕ) (x y : ℕ → ℝ) (σ : equiv.perm ℕ) (hx : antitone_on x (finset.Icc 1 n))
+  (hy : antitone_on y (finset.Icc 1 n)) (hσ : {x | σ x ≠ x} ⊆ finset.Icc 1 n)
+include hx hy hσ
+
+theorem IMO_1975_Q1 :
+  ∑ i in finset.Icc 1 n, (x i - y i) ^ 2 ≤ ∑ i in finset.Icc 1 n, (x i - y (σ i)) ^ 2 :=
+begin
+  simp only [sub_sq, finset.sum_add_distrib, finset.sum_sub_distrib],
+  -- a finite sum is invariant if we permute the order of summation
+  have hσy : ∑ (i : ℕ) in finset.Icc 1 n, y i ^ 2 = ∑ (i : ℕ) in finset.Icc 1 n, y (σ i) ^ 2,
+  { conv_lhs { rw ← finset.map_perm hσ },
+    rw finset.sum_map (finset.Icc 1 n) (σ : ℕ ↪ ℕ) _,
+    congr },
```

Comment:
Can't you use `equiv.perm.sum_comp` here?

Reply:
I can! Golfed now

Reply:
Can you even inline the `have` or does it not unify?

Reply:
I left it because it flows nicely with my comments but I could delete it. I thought that with IMO solutions, one should also care about clarity of the steps taken